repos/system_upgrade/common/actors/peseventsscanner/libraries/pes_events_scanner.py

```diff
@@ -163,6 +170,55 @@ def compute_pkg_changes_between_consequent_releases(source_installed_pkgs,
     return (target_pkgs, pkgs_to_demodularize)
 
 
+def remove_undesired_events(events, relevant_to_releases):
+    """
+    Conservatively remove events that needless, or cause problems for the current implementation:
+    - (needless) events with to_release not in relevant releases
+    - (problematic) events with the same from_release and the same in_pkgs
+    """
+
+    logger = api.current_logger()
+    relevant_to_releases = set(relevant_to_releases)
+
+    events_with_same_in_pkgs_and_from_release = defaultdict(list)
+    for event in events:
+        if event.to_release in relevant_to_releases:
+            # NOTE(mhecko): The tuple(sorted(event.in_pkgs))) is ugly, however, the removal of the events with the same
+            # #             from_release and in_pkgs is needed only because the current implementation is flawed.
+            # #             I would love to rewrite the core algorithm as a "solution to graph reachability problem",
+            # #             making the behaviour of PES event scanner purely data driven.
+            events_with_same_in_pkgs_and_from_release[(event.from_release, tuple(sorted(event.in_pkgs)))].append(event)
+
+    cleaned_events = []
+    for from_release_in_pkgs_pair, problematic_events in events_with_same_in_pkgs_and_from_release.items():
+        if len(problematic_events) == 1:
+            cleaned_events.append(problematic_events[0])  # There is no problem
+            continue
+
+        # E.g., one of the problematic events is to=8.6, other one to=8.7, keep only 8.7
+        from_release, dummy_in_pkgs = from_release_in_pkgs_pair
```

Comment:
@MichalHe  i am ok with it, but in this case I would not even create `dummy_in_pkgs` var as it's clear what is the second item in the tuple regarding the name, and usually this (`x, dummy_y = z`) way is better when you want to clearly specify what is the other item for a possible later use. so no problem with it at all, just wanted to give you a tip. keep it as it is now. most likely we are going to finish the merge in upcoming minutes :) i am jut finishing my "painting" of graphs